public/components/custom_panels/redux/panel_slice.ts

```diff
@@ -112,34 +113,47 @@ export const fetchPanel = (id) => async (dispatch, getState) => {
 
 export const fetchVisualization = () => (dispatch, getState) => {};
 
-const updateLegacyPanel = (panel: CustomPanelType) => coreRefs.http!
-  .post(`${CUSTOM_PANELS_API_PREFIX}/panels/update`, {
+const updateLegacyPanel = (panel: CustomPanelType) =>
+  coreRefs.http!.post(`${CUSTOM_PANELS_API_PREFIX}/panels/update`, {
     body: JSON.stringify({ panelId: panel.id, panel: panel as PanelType }),
   });
 
 const updateSavedObjectPanel = (panel: CustomPanelType) => savedObjectPanelsClient.update(panel);
 
-
-const uuidRx = /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/;
+export const uuidRx = /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/;
 
 const isUuid = (id) => !!id.match(uuidRx);
 
-
 export const updatePanel = (panel: CustomPanelType) => async (dispatch, getState) => {
   try {
-    if (isUuid(panel.id))
-      await updateSavedObjectPanel(panel)
-    else
-      await updateLegacyPanel(panel)
+    if (isUuid(panel.id)) await updateSavedObjectPanel(panel);
+    else await updateLegacyPanel(panel);
 
     dispatch(setPanel(panel));
     const panelList = getState().customPanel.panelList.map((p) => (p.id === panel.id ? panel : p));
     dispatch(setPanelList(panelList));
   } catch (err) {
-    console.log("Error updating panel", { err, panel })
+    console.log('Error updating panel', { err, panel });
   }
 };
 
+export const addVizToPanels = (panels, vizId) => async (dispatch, getState) => {
+  forEach(panels, (oldPanel) => {
+    const panel = getState().customPanel.panelList.find((p) => p.id === oldPanel.panel.id);
+
+    const allVisualizations = panel!.visualizations;
+
+    const visualizationsWithNewPanel = addVisualizationPanel(vizId, undefined, allVisualizations);
```

Comment:
@ps48 @joshuali925 can you verify this line makes sense? Since we want to add a viz to panel as a new one, pass in undefined into this function: https://github.com/opensearch-project/dashboards-observability/blob/5f2b777a2c7963680f4bea3746e6badea65b9303/public/components/custom_panels/helpers/add_visualization_helper.ts#L57?

Reply:
Yes, this is correct for adding panels to a visualization. For replace workflow, you can use the same function and pass OldVisualizationId.